### PR TITLE
Add WaveVisual system

### PR DESCRIPTION
**Pull request by iche033**
Depends on https://github.com/ignitionrobotics/ign-gazebo/pull/1275

Added WaveVisual system for rendering waves.

I tweaked the <gain> param slightly to better see if the rendering is in sync with the physics side. This is testing with a gain value of 0.5:

usv_waves

There is a bit of rocking motion in the USV but looks like rendering is in sync with physics. I also tried with higher wave gains and noticed that the boat becomes more unstable and will be pushed higher above the water surface (and lower into the water). We'll likely need to address this later.

For now, it's set at a gain of 0.3 which produces amplitudes of `0.0159463, 0.0819574, 0.0537983` - this is roughly [sea state 1](https://en.wikipedia.org/wiki/Sea_state).

I have also updated the [Vessel A](https://app.ignitionrobotics.org/OpenRobotics/fuel/models/Vessel%20A) model on Fuel. The model's `Surface` plugin is removed from the model.sdf and moved to the world file so that it's easier to add and update wavefield params. Make sure to remove your fuel cache of this model when testing.

### CODE REVIEW

File: mbzirc_ign/models/waves/materials/programs/GerstnerWaves_fs_330.glsl [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  // Gazebo requires rotated cube map lookup.
+  // R = vec3(R.x, R.z, R.y);
+
+  // Get environment color of reflected ray:
+  vec4 envColor = texture(cubeMap, R, 0.0);
+
+  // Cheap hdr effect:
+  envColor.rgb *= (envColor.r+envColor.g+envColor.b)*hdrMultiplier;
+
+  // Compute refraction ratio (Fresnel):
```
Comment (caguero) [inline mbzirc_ign/models/waves/materials/programs/GerstnerWaves_fs_330.glsl:62]: Are we doing reflection/refraction like in the VRX visual plugin?

File: mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+//
+// The derivative terms (Tangent, Binormal, Normal) have been
+// updated to be consistent with this convention.
+//
+
+#version 330
+
+in vec4 vertex;
+in vec4 uv0;
+uniform mat4 worldviewproj_matrix;
```
Comment (caguero) [inline mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl:41]: nit: camelCase?

File: mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+/////////// Input parameters //////////
+// Waves
+uniform int Nwaves;
+uniform vec3 camera_position_object_space;
```
Comment (caguero) [inline mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl:46]: nit: camelCase?

File: mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  // Compute (Surf2World * Rescale) matrix
+  B = normalize(B)*rescale;
+  T = normalize(T)*rescale;
+  N = normalize(N);
+  outVs.rotMatrix = mat3(B, T, N);
+
+  gl_Position = worldviewproj_matrix * P;
+  // gl_Position = worldviewproj_matrix * vertex;
```
Comment (caguero) [inline mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl:142]: Remove?

File: mbzirc_ign/src/WaveVisual.cc [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+        _ecm.ComponentData<components::SourceFilePath>(modelEntity).value();
+  }
+
+  // parse path to shaders
+  sdf::ElementPtr shaderElem = sdf->GetElement("shader");
+  if (!shaderElem->HasElement("vertex") ||
+      !shaderElem->HasElement("fragment"))
+  {
+    ignerr << "<shader> must have <vertex> and <fragment> sdf elements"
+           << std::endl;
```
Comment (caguero) [inline mbzirc_ign/src/WaveVisual.cc:139]: Do you need to return here?

File: mbzirc_ign/src/WaveVisual.cc [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+{
+  IGN_PROFILE("WaveVisual::PreUpdate");
+  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  this->dataPtr->currentSimTime = _info.simTime;
+}
+
+//////////////////////////////////////////////////
+void WaveVisualPrivate::OnUpdate()
+{
+  std::lock_guard<std::mutex> lock(this->mutex);
```
Comment (caguero) [inline mbzirc_ign/src/WaveVisual.cc:176]: Should we move this `lock_guard` only around the place where we read `currentSimTime`?

File: mbzirc_ign/src/WaveVisual.cc [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+
+  // params that only need to be set once
+  if (!this->paramsSet)
+  {
+    auto vsParams = this->material->VertexShaderParams();
+
+    (*vsParams)["worldviewproj_matrix"] = 1;
+
+    (*vsParams)["Nwaves"] = static_cast<int>(this->wavefield.Number());
+    (*vsParams)["rescale"] = 0.5f;
```
Comment (caguero) [inline mbzirc_ign/src/WaveVisual.cc:240]: Is this mapped to `wavefield.Scale()`?

File: mbzirc_ign/models/waves/materials/programs/GerstnerWaves_fs_330.glsl (hunk 1 above)
Comment (iche033) [reply mbzirc_ign/models/waves/materials/programs/GerstnerWaves_fs_330.glsl:62]: This is the simple version before the VRX visual plugin was implemented. Refraction is done based on wave normals and view angle, and only the sky in visible in the reflection (no dynamic objects).

File: mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl (hunk 2 above)
Comment (iche033) [reply mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl:43]: oh this has to map to a predefined shader param constant (defined by ogre). I've added a comment. 6b1390777afb53df8a9011cefd6b4649eb19baf9

File: mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl (hunk 3 above)
Comment (iche033) [reply mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl]: same as previous comment about mapping to predefined constant.

File: mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl (hunk 4 above)
Comment (iche033) [reply mbzirc_ign/models/waves/materials/programs/GerstnerWaves_vs_330.glsl]: removed. 6b1390777afb53df8a9011cefd6b4649eb19baf9

File: mbzirc_ign/src/WaveVisual.cc (hunk 5 above)
Comment (iche033) [reply mbzirc_ign/src/WaveVisual.cc:139]: yep done. 6b1390777afb53df8a9011cefd6b4649eb19baf9

File: mbzirc_ign/src/WaveVisual.cc (hunk 6 above)
Comment (iche033) [reply mbzirc_ign/src/WaveVisual.cc]: moved. 6b1390777afb53df8a9011cefd6b4649eb19baf9

File: mbzirc_ign/src/WaveVisual.cc (hunk 7 above)
Comment (iche033) [reply mbzirc_ign/src/WaveVisual.cc:241]: this is for scaling the normals in the bump map - affects visual appearance only. Added more comments on the params in 6b1390777afb53df8a9011cefd6b4649eb19baf9

**Review by caguero — APPROVED** [review]

**Comment by iche033** [comment]
I'll do an ign-gazebo6 release before merging this PR

**Pull request merged**